source/server/listener_manager_impl.cc

```diff
@@ -537,6 +580,26 @@ bool ListenerManagerImpl::addOrUpdateListener(const envoy::api::v2::Listener& co
   } else {
     name = server_.random().uuid();
   }
+  auto it = error_state_tracker_.find(name);
+  try {
+    return addOrUpdateListenerInternal(config, version_info, added_via_api, name);
+  } catch (const EnvoyException& e) {
+    if (it == error_state_tracker_.end()) {
+      it = error_state_tracker_.emplace(name, std::make_unique<UpdateFailureState>()).first;
+    }
+    TimestampUtil::systemClockToTimestamp(server_.api().timeSource().systemTime(),
+                                          *(it->second->mutable_last_update_attempt()));
+    it->second->set_details(e.what());
+    it->second->mutable_failed_configuration()->PackFrom(config);
```

Comment:
I would probably approach this as adding the per-listener error state to the exception (via a move) and then recovering it in the outer catch above, to avoid tangling the per-listener update code with listener manager wide error state. But, there are many ways to skin a cat :)